pkg/ride/tree_evaluator.go

```diff
@@ -285,40 +399,86 @@ func (e *treeEvaluator) evaluateNativeFunction(name string, arguments []Node) (r
 		return nil, EvaluationFailure.Errorf("failed to get cost of system function '%s'", name)
 	}
 	defer func() {
-		e.complexity += cost
+		e.cc.addNativeFunctionComplexity(cost)
 	}()
-	args := make([]rideType, len(arguments))
-	for i, arg := range arguments {
-		a, err := e.walk(arg) // materialize argument
-		if err != nil {
-			return nil, EvaluationErrorPush(err, "failed to materialize argument %d of system function '%s'", i+1, name)
-		}
-		args[i] = a
+	args, err := e.materializeArguments(arguments)
+	if err != nil {
+		return nil, EvaluationErrorPush(err, "failed to call system function '%s'", name)
 	}
-	r, err := f(e.env, args...)
+	r, err := f(e, e.env, args...)
 	if err != nil {
 		return nil, EvaluationErrorPush(err, "failed to call system function '%s'", name)
 	}
 	return r, nil
 }
 
+func (e *treeEvaluator) evaluateUserFunction(name string, args []rideType) (rideType, error) {
+	initialComplexity := e.cc.complexity()
+	defer func() {
+		if initialComplexity == e.cc.complexity() {
+			e.cc.addAdditionalUserFunctionComplexity()
+		}
+	}()
+	uf, cl, found := e.s.userFunction(name)
+	if !found {
+		return nil, errors.Errorf("user function '%s' not found", name)
+	}
+	if len(args) != len(uf.Arguments) {
+		return nil, errors.Errorf("mismatched arguments number of user function '%s'", name)
+	}
+	avs := make([]esValue, len(args))
+	for i, arg := range args {
+		an := uf.Arguments[i]
+		avs[i] = esValue{id: an, value: arg}
+	}
+	e.s.cs = append(e.s.cs, make([]esValue, len(args)))
+	for i, av := range avs {
+		e.s.cs[len(e.s.cs)-1][i] = av
+	}
```

Comment:
Why you can't do it like that
```go
	e.s.cs = append(e.s.cs, avs)
```
instead of
```go
	e.s.cs = append(e.s.cs, make([]esValue, len(args)))
	for i, av := range avs {
		e.s.cs[len(e.s.cs)-1][i] = av
	}
```
?

Reply:
May be it's because there was no `avs` before.